File: cs/Parking/Manager/ParkingManager.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Parking
{
    public class ParkingManager
    {
        private readonly IReadOnlyList<ParkingBoy> parkingBoys;

        public ParkingManager(IReadOnlyList<ParkingBoy> parkingBoys)
        {
            this.parkingBoys = parkingBoys;
        }

        private ParkingBoy FindOneValidParkingBoy() =>
        parkingBoys
        .OrderBy(_ => new Random().Next(0, parkingBoys.Count))
        .FirstOrDefault(parkingBoy => parkingBoy.CheckAvailableLot());

        public ParkingLot FindOneValidParkingLot() =>
            FindOneValidParkingBoy()?.FindOneValidParkingLot();
    }
}

File: ts/src/parking/domain/parkingLot/finder/parkingManager.ts
/** @format */

import {ParkingBoy} from './parkingBoy'
import {ParkingLotFinder} from '../parking/parkingLotFinder'

export class ParkingManager implements ParkingLotFinder {
    public constructor(private parkingBoys: ParkingBoy[]) {}

    private findOneValidParkingBoy() {
        return this.parkingBoys
            .sort(() => Math.floor(Math.random() * this.parkingBoys.length))
            .find(parkingBoy => parkingBoy.checkAvailableLot())
    }

    public findOneValidParkingLot() {
        const parkingBoy = this.findOneValidParkingBoy()
        return parkingBoy?.findOneValidParkingLot() ?? null
    }
}
